assets/javascripts/test_result.js

```diff
@@ -778,60 +793,124 @@ function renderInvestigationTab(response) {
-        html.innerHTML = value.text;
-        valueElement.appendChild(html);
-      }
-    } else {
-      var textLines = typeof value === 'string' ? value.split('\n') : [value];
-      var textLinesRest;
-
-      var lineLimit = 10;
-      if (textLines.length > lineLimit) {
-        textLinesRest = textLines.slice(lineLimit, textLines.length);
-        textLines = textLines.slice(0, lineLimit);
-      }
-
-      var preElement = document.createElement('pre');
-      preElement.appendChild(document.createTextNode(textLines.join('\n')));
-      valueElement.appendChild(preElement);
-
-      if (textLinesRest) {
-        var preElementMore = document.createElement('pre');
-        preElementMore.appendChild(document.createTextNode(textLinesRest.join('\n')));
-        preElementMore.style = 'display: none';
-
-        var moreLink = document.createElement('a');
-        moreLink.style = 'cursor:pointer';
-        moreLink.innerHTML = 'Show more';
-        moreLink.onclick = function () {
-          preElementMore.style = '';
-          moreLink.style = 'display:none';
-        };
-
-        valueElement.appendChild(moreLink);
-        valueElement.appendChild(preElementMore);
-      }
     }
 
-    var trElement = document.createElement('tr');
-    trElement.appendChild(keyElement);
-    trElement.appendChild(valueElement);
-    tbodyElement.appendChild(trElement);
-  });
+    var theadElement = document.createElement('thead');
+    var headTrElement = document.createElement('tr');
+    var headThElement = document.createElement('th');
+    headThElement.appendChild(document.createTextNode('Investigation'));
+    headThElement.colSpan = 2;
+    headTrElement.appendChild(headThElement);
+    theadElement.appendChild(headTrElement);
+
+    var tbodyElement = document.createElement('tbody');
+    var alertbox;
+    Object.keys(response).forEach(key => {
+        var value = response[key];
+        var type = 'pre';
+
+        // The value can be an object with attribute "type" to determine the
+        // behavior. The accepted types are:
+        // - link: adds a link reference using an anchor <a>
+        //   additional required attributes:
+        //     - link: the url
+        //     - text: the text to show instead of the url
+
+        if (typeof value === 'object' && value.type)
+            type = value.type;
+
+        var keyElement = document.createElement('td');
+        keyElement.style.verticalAlign = 'top';
+        keyElement.appendChild(document.createTextNode(key));
+
+        var valueElement = document.createElement('td');
+
+        if (type === "link") {
+            var html = document.createElement("a");
+            if (key === 'first_bad') {
+                alertbox = document.createElement("div");
+                alertbox.appendChild(document.createTextNode('Investigate the first bad test directly: '));
+                html.href = value.link + '#investigation';
+                html.innerHTML = value.text;
```

Comment:
Does this need escaping? Maybe you can just append a text node as child here?

Reply:
Actually i havent touched this part at all. this block is never parsed against the `test_log`. i am not sure if i miss something on your comment.

Reply:
Ok, if it is just a reformatting/move of existing code, leave it as-is. I've just noticed it within the diff. (It could be that `value.text` is actually supposed to contain HTML-syntax here.)